site/content/blog/2020-03-16-advanced-git-tbilisi-free-university.md

```diff
@@ -0,0 +1,437 @@
+
+![image](/blog/2020-03-16-advanced-git-tbilisi-free-university/branching-overview.png)
+
+Let's look at the most recent such feature release: v5.20.
+
+```sh
+# Checkout the mattermost-server repository (if not already done)
+git checkout https://github.com/mattermost/mattermost-server.git
+cd mattermost-server
+
+# Examine the combined history of release-5.20 and master
+git log --graph --oneline release-5.20 master
+```
+
+Observe that `master` has advanced in the weeks since we released v5.20. When we first released v5.20.0, a community member tried installing it and promptly [reported](https://community.mattermost.com/core/pl/877mkcc967ytjpgcgpbo6bi7ty) that his server was crashing! A bit of debugging on a Saturday, and I realized that we had introduced a regression into a code path we didn't often use ourselves. I filed [MM-21619](https://mattermost.atlassian.net/browse/MM-22619) and got to work on the fix. Let's find it in the `master` branch.
+
+```sh
+# Find the commit for MM-22619 in master and release 5.20
+git log --grep MM-22619 master
+```
+
+Now let's try to repeat that cherry-pick from the original v5.20.0 tag. 
+
+```sh
+git checkout -b test-git-cherrypick v5.20.0
+git cherry-pick 9a51c73f6428b70e31fc8c35de770b91270e6bba
+```
+
+Oh no, merge conflicts! Two of the files were cherry-picked without issue, but some other changes had snuck into master before my fix. Whenever a merge conflict occurs, Git annotates the file with conflict markers. Let's open up the file and take a look. There are three blocks: HEAD (what's in my current branch), parent (the common ancestor of the cherry-pick and this branch), and the cherry-pick itself.
+
+```patch
+<<<<<<< HEAD
+	th := Setup(t).InitBasic()
+||||||| parent of 9a51c73f6... MM-22619: check for nil plugins environment (#13901)
+	th := Setup(t)
+=======
+	th := SetupWithCustomConfig(t, func(config *model.Config) {
+		*config.PluginSettings.Enable = false
+	})
+>>>>>>> 9a51c73f6... MM-22619: check for nil plugins environment (#13901)
+	defer th.TearDown()
+```
+
+I have a mergetool configured with Git to show this same information using a three-way merge inside Vim:
+
+```sh
+git mergetool
+```
+
+The middle pane shows the common ancestor. On its left is what's in the current branch, and on the right is what's incoming. The bottom pane is the current file, showing the conflict markers. In general with merge conflicts, you have to understand the context of the code you are changing in order to make the decision on what version to keep, or whether to blend the changes somehow. In this case, I happen to know that a colleague of mine removed a redundant call to `InitBasic` in the master branch as part of a cleanup effort, but those clean up changes aren't in this release branch. So I need to combine the two changes together and put the call to `InitBasic` on top of the my original code. Then I'll get rid of the other sections and the conflict markers.
```

Comment:
Nit: I think that a screenshot of your editor with what you're describing here would be terrific! (and it's always cool to look at the editors of other developers!)

Reply:
Good call! Let me get that re-setup.